File: MiniDOMWebLibs/MiniDOM/Session/Sessions.cs
using System;
using DMD;
using DMD.XML;
using minidom;
using minidom.Internals;
using static minidom.Sistema;

namespace minidom
{
    namespace Internals
    {

        [Serializable]
        public sealed class CSessionsClass 
            : Sistema.CModulesClass<WebSite.CSiteSession>
        {

            /// <summary>
        /// Evento generato quando viene iniziata una nuova sessione remota
        /// </summary>
        /// <param name="sender"></param>
        /// <param name="e"></param>
            public event SessionStartEventHandler SessionStart;

            public delegate void SessionStartEventHandler(object sender, WebSite.SessionEventArgs e);


            /// <summary>
            /// Evento generato quando viene terminata una sessione remota
            /// </summary>
            /// <param name="sender"></param>
            /// <param name="e"></param>
            public event SessionEndEventHandler SessionEnd;

            public delegate void SessionEndEventHandler(object sender, WebSite.SessionEventArgs e);

            private int m_ActiveSessions = 0;

            internal CSessionsClass() : base("modSiteSessions", typeof(WebSite.CSiteSessionsCursor), 0)
            {
            }

            public int CountActiveSessions()
            {
                return m_ActiveSessions;
            }

            public void NotifySessionStart(WebSite.SessionEventArgs e)
            {
                m_ActiveSessions += 1;
                SessionStart?.Invoke(null, e);
            }

            public void NotifySessionEnd(WebSite.SessionEventArgs e)
            {
                m_ActiveSessions -= 1;
                SessionEnd?.Invoke(null, e);
            }
        }
    }

    public partial class WebSite
    {
        private static CSessionsClass m_Sessions = null;

        public static CSessionsClass Sessions
        {
            get
            {
                if (m_Sessions is null)
                    m_Sessions = new CSessionsClass();
                return m_Sessions;
            }
        }
    }
}

File: MiniDOMWebLibs/MiniDOM/Session/Sessions.vb
Imports minidom
Imports minidom.Databases
Imports minidom.Sistema
Imports minidom.Internals
Imports minidom.WebSite

Namespace Internals

    Public NotInheritable Class CSessionsClass
        Inherits CModulesClass(Of CSiteSession)

        ''' <summary>
        ''' Evento generato quando viene iniziata una nuova sessione remota
        ''' </summary>
        ''' <param name="sender"></param>
        ''' <param name="e"></param>
        Public Event SessionStart(ByVal sender As Object, ByVal e As SessionEventArgs)


        ''' <summary>
        ''' Evento generato quando viene terminata una sessione remota
        ''' </summary>
        ''' <param name="sender"></param>
        ''' <param name="e"></param>
        Public Event SessionEnd(ByVal sender As Object, ByVal e As SessionEventArgs)


        Private m_ActiveSessions As Integer = 0


        Friend Sub New()
            MyBase.New("modSiteSessions", GetType(CSiteSessionsCursor), 0)
        End Sub



        Public Function CountActiveSessions() As Integer
            Return m_ActiveSessions
        End Function

        Public Sub NotifySessionStart(ByVal e As SessionEventArgs)
            m_ActiveSessions += 1
            RaiseEvent SessionStart(Nothing, e)
        End Sub


        Public Sub NotifySessionEnd(ByVal e As SessionEventArgs)
            m_ActiveSessions -= 1
            RaiseEvent SessionEnd(Nothing, e)
        End Sub

    End Class
End Namespace


Partial Class WebSite



    Private Shared m_Sessions As CSessionsClass = Nothing

    Public Shared ReadOnly Property Sessions As CSessionsClass
        Get
            If (m_Sessions Is Nothing) Then m_Sessions = New CSessionsClass
            Return m_Sessions
        End Get
    End Property

End Class
